cores/esp8266/Esp.cpp

```diff
@@ -692,39 +698,141 @@ static SpiFlashOpResult spi_flash_write_puya(uint32_t offset, uint32_t *data, si
-        rc = spi_flash_write_puya(offset, data, size);
+        rc = spi_flash_write_puya(offset, const_cast<uint32_t *>(data), size);
     }
     else
-#endif
+#endif // PUYA_SUPPORT
     {
-        rc = spi_flash_write(offset, data, size);
+        rc = spi_flash_write(offset, const_cast<uint32_t *>(data), size);
     }
     return rc == SPI_FLASH_RESULT_OK;
 }
 
+static const int UNALIGNED_WRITE_BUFFER_SIZE = 512;
+
+bool EspClass::flashWrite(uint32_t offset, const uint8_t *data, size_t size) {
+    size_t sizeAligned = size & ~3;
+    size_t currentOffset = 0;
+    if ((uintptr_t)data % 4 != 0) {
+        // Memory is unaligned, so we need to copy it to an aligned buffer
+        uint32_t alignedData[UNALIGNED_WRITE_BUFFER_SIZE / sizeof(uint32_t)] __attribute__((aligned(4)));
+        size_t sizeLeft = sizeAligned;
+
+        while (sizeLeft) {
+            size_t willCopy = std::min(sizeLeft, sizeof(alignedData));
+            memcpy(alignedData, data + currentOffset, willCopy);
+            // We now have data and size aligned to 4 bytes, so we can use aligned write
+            if (!flashWrite(offset + currentOffset, alignedData, willCopy))
+            {
+                return false;
+            }
+            sizeLeft -= willCopy;
+            currentOffset += willCopy;
+        }
+    } else {
+        // Pointer is properly aligned, so use aligned write
+        if (!flashWrite(offset, (uint32_t *)data, sizeAligned)) {
+            return false;
+        }
+        currentOffset = sizeAligned;
+    }
+    if (currentOffset < size) {
+        // Size was not aligned, but writes must be 4-byte aligned
+#if PUYA_SUPPORT
+        if (getFlashChipVendorId() == SPI_FLASH_VENDOR_PUYA) {
+            uint32_t tempData;
+            if (spi_flash_read(offset + currentOffset, &tempData, 4) != SPI_FLASH_RESULT_OK) {
+                return false;
+            }
+            for (size_t i = 0; i < size - currentOffset; i++) {
+                ((uint8_t *)tempData)[i] &= ((uint8_t *)data)[currentOffset + i];
+            }
+            if (spi_flash_write(currentOffset, &tempData, 4) != SPI_FLASH_RESULT_OK) {
+                return false;
+            }
+        }
+        else
+#endif // PUYA_SUPPORT
+        {
+            uint32_t tempData;
+            if (spi_flash_read(offset + currentOffset, &tempData, 4) != SPI_FLASH_RESULT_OK) {
+                return false;
+            }
+            memcpy(&tempData, (uint8_t *)data + currentOffset, size - currentOffset);
+            if (spi_flash_write(offset + currentOffset, &tempData, 4) != SPI_FLASH_RESULT_OK) {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+bool EspClass::flashRead(uint32_t offset, uint8_t *data, size_t size) {
+    size_t sizeAligned = size & ~3;
+    size_t currentOffset = 0;
+
+    if ((uintptr_t)data % 4 != 0) {
+        uint32_t alignedData[UNALIGNED_WRITE_BUFFER_SIZE / sizeof(uint32_t)] __attribute__((aligned(4)));
```

Comment:
This uses up 512 bytes on the stack, which is just 4KB. I'm not sure whether that's ok, but we're trying to not have large stack usage.

Reply:
From the comment in `flash_hal.c`:
>The size of this buffer is a tradeoff between number of writes required
 and amount of stack required. This is chosen to be 512 bytes here, but might
 be adjusted in the future if there are good reasons to do so.
I think I can lower it to 256 (page size), plus it only happens when unaligned memory is passed